packages/framework/src/Controller/Admin/MailController.php

```diff
@@ -183,4 +194,56 @@ public function settingAction(Request $request): Response
             'form' => $form->createView(),
         ]);
     }
+
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     * @param int $mailTemplateId
+     * @return \Symfony\Component\HttpFoundation\Response
+     */
+    #[Route(path: '/mail/send/{mailTemplateId}', requirements: ['mailTemplateId' => '\d+'])]
+    public function sendAction(Request $request, int $mailTemplateId): Response
+    {
+        $mailTemplate = $this->mailTemplateFacade->getById($mailTemplateId);
+        $form = $this->createForm(MailTemplateSendFormType::class, null, [
+            'action' => $this->generateUrl('admin_mail_send', ['mailTemplateId' => $mailTemplateId]),
+            'mailTemplate' => $mailTemplate,
+        ]);
+        $form->handleRequest($request);
+
+        if ($form->isSubmitted() && $form->isValid()) {
+            $data = $form->getData();
+
+            try {
+                $this->mailTemplateSenderFacade->sendMail($mailTemplate, $data['mailTo'], $data['prefillFromEntityIdentifier'] ?? null);
+                $this->addSuccessFlash(t('Test email sent to %email%.', ['%email%' => $data['mailTo']]));
+
+                return new JsonResponse(['result' => 'valid']);
+            } catch (Throwable $exception) {
+                $this->addErrorFlash(t('Error occurred while sending email.'));
+                $this->addErrorFlash($exception->getMessage());
+
+                return $this->getInvalidResponse($form);
+            }
+        }
+
+        if ($form->isSubmitted() && !$form->isValid()) {
+            return $this->getInvalidResponse($form);
+        }
+
+        return $this->render('@ShopsysFramework/Admin/Content/Mail/mailTemplateSend.html.twig', [
+            'form' => $form->createView(),
+        ]);
+    }
+
+    /**
+     * @param \Symfony\Component\Form\FormInterface $form
+     * @return \Symfony\Component\HttpFoundation\JsonResponse
+     */
+    protected function getInvalidResponse(FormInterface $form): JsonResponse
```

Comment:
maybe `createInvalidResponse` as the method always creates a new instance

Reply:
:+1: fixed in 7baa7349eb